src/dbally/iql_generator/prompt.py

```diff
@@ -71,82 +112,131 @@ def __init__(self, *, question: str, examples: List[FewShotExample] = None) -> N
-        examples: List[FewShotExample] = None,
+        methods: List[ExposedFunction],
+        examples: Optional[List[FewShotExample]] = None,
     ) -> None:
         """
         Constructs a new IQLGenerationPromptFormat instance.
 
         Args:
             question: Question to be asked.
-            filters: List of filters exposed by the view.
+            methods: List of filters exposed by the view.
             examples: List of examples to be injected into the conversation.
+            aggregations: List of aggregations exposed by the view.
         """
         super().__init__(examples)
         self.question = question
-        self.filters = "\n".join([str(filter) for filter in filters])
+        self.methods = "\n".join([str(condition) for condition in methods]) if methods else []
 
 
-IQL_GENERATION_TEMPLATE = PromptTemplate[IQLGenerationPromptFormat](
+FILTERING_DECISION_TEMPLATE = PromptTemplate[DecisionPromptFormat](
     [
         {
             "role": "system",
             "content": (
-                "You have access to API that lets you query a database:\n"
-                "\n{filters}\n"
-                "Please suggest which one(s) to call and how they should be joined with logic operators (AND, OR, NOT).\n"
-                "Remember! Don't give any comments, just the function calls.\n"
-                "The output will look like this:\n"
-                'filter1("arg1") AND (NOT filter2(120) OR filter3(True))\n'
-                "DO NOT INCLUDE arguments names in your response. Only the values.\n"
-                "You MUST use only these methods:\n"
-                "\n{filters}\n"
-                "It is VERY IMPORTANT not to use methods other than those listed above."
-                """If you DON'T KNOW HOW TO ANSWER DON'T SAY \"\", SAY: `UNSUPPORTED QUERY` INSTEAD! """
-                "This is CRUCIAL, otherwise the system will crash. "
+                "Given a question, determine whether the answer requires initial data filtering in order to compute it.\n"
+                "Initial data filtering is a process in which the result set is reduced to only include the rows "
+                "that meet certain criteria specified in the question.\n\n"
+                "---\n\n"
+                "Follow the following format.\n\n"
+                "Question: ${{question}}\n"
+                "Hint: ${{hint}}"
+                "Reasoning: Let's think step by step in order to ${{produce the decision}}. We...\n"
+                "Decision: indicates whether the answer to the question requires initial data filtering. "
+                "(Respond with True or False)\n\n"
             ),
         },
         {
             "role": "user",
-            "content": "{question}",
+            "content": (
+                "Question: {question}\n"
+                "Hint: Look for words indicating data specific features.\n"
+                "Reasoning: Let's think step by step in order to "
```

Comment:
It's not clear to me why `Hint` and `Reasoning` are repeated in System and User message, although in System they look dynamic (thought I'm not sure where the values come from) and in User they look static.

Reply:
We use system prompt to define a task and a structure of the conversation. Some parts of the prompts are static, some are dynamic. Here we tell the LLM to follow the format and we help it maintain the structure in the user prompt, so the LLM knows how to generate reasoning based on the hint we have already provided. There is only one user input attribute and it is `{question}`, the double brackets `{{}}` indicate that this is an attribute for the LLM to fill in, the hint is already provided, so the LLM will start generating with a fixed hint and therefore should achieve better accuracy. 

I've tested this on our benchmark and prompts with hints on CoT performs better in general on smaller models, since they require more structure.